cdc/sinkv2/ddlsink/mysql/mysql_ddl_sink.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"net/url"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	timodel "github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tiflow/cdc/contextutil"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sinkv2/ddlsink"
+	"github.com/pingcap/tiflow/cdc/sinkv2/metrics"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/errorutil"
+	"github.com/pingcap/tiflow/pkg/quotes"
+	"github.com/pingcap/tiflow/pkg/retry"
+	"github.com/pingcap/tiflow/pkg/sink"
+	pmysql "github.com/pingcap/tiflow/pkg/sink/mysql"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultDDLMaxRetry uint64 = 20
+)
+
+// Assert DDLEventSink implementation
+var _ ddlsink.DDLEventSink = (*mysqlDDLSink)(nil)
+
+type mysqlDDLSink struct {
+	// id indicates which processor (changefeed) this sink belongs to.
+	id model.ChangeFeedID
+	// db is the database connection.
+	db *sql.DB
+	// statistics is the statistics of this sink.
+	// We use it to record the DDL count.
+	statistics *metrics.Statistics
+}
+
+// NewMySQLDDLSink creates a new mysqlDDLSink.
+func NewMySQLDDLSink(
+	ctx context.Context,
+	sinkURI *url.URL,
+	replicaConfig *config.ReplicaConfig,
+	dbConnFactory pmysql.Factory,
+) (*mysqlDDLSink, error) {
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
+	cfg := pmysql.NewConfig()
+	err := cfg.Apply(ctx, changefeedID, sinkURI, replicaConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	dsnStr, err := pmysql.GenerateDSN(ctx, sinkURI, cfg, dbConnFactory)
+	if err != nil {
+		return nil, err
+	}
+
+	db, err := dbConnFactory(ctx, dsnStr)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &mysqlDDLSink{
+		id:         changefeedID,
+		db:         db,
+		statistics: metrics.NewStatistics(ctx, sink.TxnSink),
```

Comment:
It's better to use a sub-context `ctx1, cancel = context.WithCancel(ctx)`, and call `cancel` when close the sink.

Reply:
It will be canceled at:
```go
func (s *ddlSinkImpl) close(ctx context.Context) (err error) {
	s.cancel()
	if s.sinkV1 != nil {
		err = s.sinkV1.Close(ctx)
	} else {
		err = s.sinkV2.Close()
	}
	if s.syncPointStore != nil {
		err = s.syncPointStore.Close()
	}
	s.wg.Wait()
	if err != nil && errors.Cause(err) != context.Canceled {
		return err
	}
	return nil
}
```